Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java

```diff
@@ -1,125 +1,110 @@
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.io.PrintWriter;
+import java.io.UncheckedIOException;
 import java.net.Socket;
 import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.ExecutionException;
-import org.jetbrains.annotations.NotNull;
 
 /** Handles every interaction of app with the server. */
 public class ServerHandler {
-
-  private Integer serverSocketPort;
-  private long serverStartTimestamp;
-  private Socket socket;
-  private BufferedReader inputReader;
-  private PrintWriter outputWriter;
-  private long hostSyncTimestamp;
-
-  public ServerHandler(Integer serverSocketPort) {
-    this.serverSocketPort = serverSocketPort;
+  private static class Connection {
+    Socket socket;
+    BufferedReader inputReader;
+    PrintWriter outputWriter;
   }
 
-  public long getHostSyncTimestamp() {
-    return hostSyncTimestamp;
+  public static class ServerData{
+    ArrayList<Long> timestamps;
+    ArrayList<String> serverSequence;
+    ServerData(){
+      this.timestamps = new ArrayList<>();
+      this.serverSequence = new ArrayList<>();
+    }
   }
 
-  public Integer getServerSocketPort() {
-    return serverSocketPort;
-  }
+  private Integer serverSocketPort;
+  private CompletableFuture<Connection> connection;
 
-  public void setServerSocketPort(Integer serverSocketPort) {
+  public ServerHandler(Integer serverSocketPort) {
     this.serverSocketPort = serverSocketPort;
   }
 
-  public long getServerStartTimestamp() {
-    return serverStartTimestamp;
-  }
-
+  /**
+   * Creates a socket for communication with the server. Sets reader and writer for server
+   * connection.
+   */
+  @RequiresApi(api = VERSION_CODES.N)
   public void startConnection() {
-    ConnectionThread connectionThread = new ConnectionThread();
-    connectionThread.execute();
+    connection =
+        CompletableFuture.supplyAsync(
+            () -> {
+              Connection conn = new Connection();
+              String SERVER_IP = "127.0.0.1";
+              try {
+                conn.socket = new Socket(SERVER_IP, serverSocketPort);
+                conn.inputReader =
+                    new BufferedReader(new InputStreamReader(conn.socket.getInputStream()));
+                conn.outputWriter = new PrintWriter(conn.socket.getOutputStream());
+              } catch (IOException e) {
+                throw new UncheckedIOException(e);
+              }
+              Log.d(ContentValues.TAG, "Connection established");
+              return conn;
+            });
   }
 
-  // TODO: Use round-trip averaging to fix hostSyncTimestamp
+  @RequiresApi(api = VERSION_CODES.N)
   public void sendKeySimulationSignal() {
-    hostSyncTimestamp = System.currentTimeMillis();
-    outputWriter.write("started capture*");
-    outputWriter.flush();
+        connection.thenApply(
```

Comment:
I think the way you are chaining futures here (and ignoring the result) means that any errors that occurred in the connection stage, or in the this stage, will never be surfaced. I think we should try to find a way for them to bubble up to the top-level control flow.